rfd/0129-discovery-name-templating.md

```diff
@@ -0,0 +1,395 @@
+#...connects to "bar-rds-us-west-2-123456789012" by matching region label...
+
+# ambiguous label(s) match is also an error
+$ tsh db connect --db-user=alice --db-name-postgres region=us-west-1 
+error: ambiguous database labels could match multiple databases:
+Name                           Description               Protocol Type URI                                                   Allowed Users Labels                                                                                                                                    Connect 
+------------------------------ ------------------------- -------- ---- ----------------------------------------------------- ------------- ----------------------------------------------------------------------------------------------------------------------------------------- ------- 
+bar-rds-us-west-1-123456789012 RDS instance in us-west-1 postgres rds  bar.abcdefghijklmnop.us-west-1.rds.amazonaws.com:5432 [*]           account-id=123456789012,endpoint-type=instance,engine-version=13.10,engine=postgres,env=dev,region=us-west-1,teleport.dev/origin=dynamic          
+foo-rds-us-west-1-123456789012 RDS instance in us-west-1 postgres rds  foo.abcdefghijklmnop.us-west-1.rds.amazonaws.com:5432 [*]           account-id=123456789012,endpoint-type=instance,engine-version=13.10,engine=postgres,env=prod,region=us-west-1,teleport.dev/origin=dynamic         
+
+Hint: try addressing the database by its full name or by matching its labels (ex: tsh db connect key1=value1,key2=value2).
+Hint: use `tsh db ls -v` or `tsh db ls --format=[yaml | json]` to list all databases with verbose details.
+
+# resolve the error by using either more specific labels or adding a prefix name
+$ tsh db connect --db-user=alice --db-name-postgres foo region=us-west-1 
+#...connects to "foo-rds-us-west-1-123456789012" by prefix and label...
+$ tsh db connect --db-user=alice --db-name-postgres region=us-west-1,env=prod
+#...connects to "foo-rds-us-west-1-123456789012" by multiple labels...
+```
+
+### Web UI and Teleport Connect UX
+
+Both the web UI and Teleport Connect already support searching for substrings
+in resource names and labels.
+
+Searching by substring is a "fuzzier" kind of search than prefix-based name
+search (like this RFD proposed prefix-based search for `tsh`) - it's more
+likely to match more than one resource.
+However, GUI UX is fundamentally different from CLI - users can search and then
+interactively select from multiple matching resources.
+So this kind of search is appropriate for the web UI and Teleport Connect, but
+not for `tsh`.
+
+Both web UI and Teleport Connect also support label-based searching with
+the predicate language, e.g.:
+
+```
+labels["env"] == "dev" && labels["region"] == "us-west-1"
+```
+
+Therefore, no UX changes are required for these user interfaces.
+
+### Security
+
+No security concerns I can think of.
+
+### Backward Compatibility
+
+No concerns I can think of.
```

Comment:
I would address here explicitly why this would be backwards compatible for existing users - because the name of auto-discovered databases will be the first component of the name so they can just keep connecting as they used to. The gotcha is that they will have to upgrade `tsh` and we should reserve this change for T14.

Reply:
You're right that if the discovery agent is updated but `tsh` isn't, then users will have to type the full name of the resource instead of just what they used before, until they upgrade `tsh` itself.

I'll add this

Reply:
Notice that currently we are storing Database Resource in our backed as [Key = datbase.Name() ](https://github.com/gravitational/teleport/blob/master/lib/services/local/databases.go#L85

 If we are going to alter the Database key by adding a discovery prefix the discovery service upgrade to v14 will lead to creation a completely different database objects and this will break the tsh v13 flow where connecting by prefix is not supported. 


> then users will have to type the full name of the resource instead of just what they used before, until they upgrade tsh itself.

This is not backward compatible and we should avoid breaking v13 tsh flow and if needed we should hide this behind feature flag. 

The other alternative is to decouple a db resource storage key for db resource name.

Reply:
The backend key will not be altered - instead of searching for the database using predicate `name == "<db>"`, I will add a function to the predicate parser for prefix name matching: `hasPrefix(name, "<db>")`.

Database discovery will construct the extra suffix detailed in this RFD in a deterministic way, and we can provide a short name by removing that suffix in tsh output. We can use the existing label `teleport.dev/origin: cloud` to identify these resources and trim the suffix for a short name display.

> This is not backward compatible and we should avoid breaking v13 tsh flow and if needed we should hide this behind feature flag.

Hide the auto-discovery renaming behind a feature flag? It seems like a minor inconvenience if this situation arises - just requires a little extra typing in `tsh` args.

Reply:
>  just requires a little extra typing in tsh args.

I see but doing this we will break tsh v13 / user automation scripts and we should and we should avoid this kind of change because it will produce a lot of support noise and S1 issues. 

@GavinFrazar @r0mant 
 Can we consider backporting  tsh prefix resolution to tsh code to v13, v12  without teleport discovery backend change ? 

This will allow to ship this feature more grateful so if a client will use up to date tsh v13 with prefix resolution after upgrading the cluster to v14 we won't break the tsh flow.

Reply:
Sure, that sounds good to me.

Reply:
3718bc7e523eca9da8e13c3dfe13ac837b403d6a